src/metabase/lib/datascript.cljc

```diff
@@ -0,0 +1,1361 @@
+   ;; `:mbql.breakout/temporal-unit` is the keyword (eg. `:month`) of the temporal unit.
+   :mbql.breakout/temporal-unit   {:db/cardinality :db.cardinality/one,  #_#_:db/valueType :db.type/keyword}
+   ;; `:mbql.breakout/binning` is a standalone entity giving the binning details.
+   :mbql.breakout/binning         component-one
+
+   :mbql.binning/num-bins         {:db/cardinality :db.cardinality/one,  #_#_:db/valueType :db.type/int}
+   :mbql.binning/bin-width        {:db/cardinality :db.cardinality/one,  #_#_:db/valueType :db.type/double}})
+
+(def ^:private schema
+  (merge metadata-schema query-schema))
+
+(def ref-attrs
+  (into #{} (for [[attr details] schema
+                  :when (= (:db/valueType details) :db.type/ref)]
+              attr)))
+
+(defn fresh-db
+  "Returns a fresh DataScript database with the schemas loaded."
+  []
+  (d/empty-db schema))
+
+;; TODO: Probably remove anything `conn`-related, and stick to DB values?
+(defn fresh-conn
+  "Create a blank mutable database with the schema but nothing else."
+  []
+  (d/conn-from-db (fresh-db)))
+
+;; ## Internal helpers
+(defn- entity? [possible-entity]
+  (datascript.impl.entity/entity? possible-entity))
+
+(defn- entities-without-uuid [db]
+  (d/q '[:find [?e ...] :in $ :where
+         [?e _ _]
+         [(missing? $ ?e :mbql/uuid)]]
+       db))
+
+(defn- add-missing-uuids [db]
+  (let [missing (entities-without-uuid db)]
+    (d/db-with db (for [e missing]
+                    [:db/add e :mbql/uuid (str (random-uuid))]))))
+
+(defn- gen-uuid [maplike]
+  (cond-> maplike
+    (not (:mbql/uuid maplike)) (assoc :mbql/uuid (str (random-uuid)))))
+
+(defn- external-key [entity]
+  (m/find-first #(get entity %)
+                [:metadata.field/id :metadata.table/id :metadata.database/id]))
+
+(defn- external-entid
+  "Given any Entity, return a portable entid for it.
+
+  For metadata entities that exist in appdb, this is eg. `[:metadata.field/id 123]`.
+  For fragments of queries (including columns), this is `[:mbql/uuid \"bcebd82a-4f29-48fa-8d8a-effd3f5dc574\"]`."
+  [entity]
+  (or (when-let [extkey (external-key entity)]
+        [extkey (get entity extkey)])
+      (when-let [the-uuid (:mbql/uuid entity)]
+        [:mbql/uuid the-uuid])
+      (throw (ex-info "Cannot generate external-entid without either an external key or UUID"
+                      {:entity (d/touch entity)}))))
+
+(defn- missing-uuids
+  "Returns those entities which don't have UUIDs *and* that needed them!
+
+  That is, things like Tables and Fields that have their own external keys are excluded.
+
+  Returns a list of Entities."
+  [db]
+  (not-empty (for [e (entities-without-uuid db)
+                   :let [entity (d/entity db e)]
+                   :when (not (external-key entity))]
+               entity)))
+
+(defn- entid? [x]
+  (and (vector? x)
+       (= (count x) 2)
+       (keyword? (first x))))
```

Comment:
Even that can confuse some random data with an entid. I would prefer something like tagged literals in EDN.

Reply:
This is only used in `clause-argument` to guess whether the (developer-supplied, in a test) input is a field ref or not. I think at that level, this amount of guesswork seems fine to me. When expression clauses are being constructed by the UI, it would have a reference to the real column already, so there's no ambiguity in production.

Reply:
I agree. It would be great to have similarly unambiguous wire/DB format too.

Reply:
Any wire format powered by DataScript would be unambiguous, because this logic is not on the deserialization path, so to speak.

Whether we're sending datoms or maps over the wire, data coming out of DataScript would be unambiguous.  What's actually stored for each argument is an entity with exactly one of:
- `:mbql/lit x` for an opaque literal `x` (which could look like an entid, though I'm not sure how that would happen)
- `:mbql/ref 7` with a literal entity ID for a column
- `:mbql/sub` 77 with a literal entity ID for a nested expression

It's only when accepting handwritten inputs in tests that there's any risk of these being confused. As I say, that seems like an acceptable risk to me.